examples/virtual_methods/pet.rs

```diff
@@ -0,0 +1,201 @@
+use glib::prelude::*;
+use glib::subclass::prelude::*;
+
+mod imp {
+    use super::*;
+
+    /// The Pet implementation struct
+    #[derive(Default)]
+    pub struct Pet {}
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for Pet {
+        const NAME: &'static str = "Pet";
+        type Type = super::Pet;
+
+        /// Override the class struct with the custom [`PetClass`]
+        type Class = super::PetClass;
+
+        /// Initialize the class struct with the default implementations of the
+        /// virtual methods.
+        fn class_init(klass: &mut Self::Class) {
+            klass.pet = Some(|obj| obj.imp().pet_default());
+            klass.feed = Some(|obj| obj.imp().feed_default());
+        }
+    }
+
+    impl ObjectImpl for Pet {}
+
+    impl Pet {
+        /// Default implementation of [`PetImpl::pet`]
+        fn pet_default(&self) -> bool {
+            // The default behaviour is unsuccessful pets
+            println!("Pet::pet_default");
+            false
+        }
+
+        /// Default implementation of [`PetImpl::feed`]
+        fn feed_default(&self) {
+            println!("Pet::feed_default");
+        }
+    }
+}
+
+glib::wrapper! {
+    /// The `Pet` class acts as a base class for pets and provides two virtual methods
+    pub struct Pet(ObjectSubclass<imp::Pet>);
+}
+
+/// Custom class struct for the [`Pet`] class.
+///
+/// The class struct is used to implement the vtable for method dispatch. The first field *must*
+/// be a pointer to the parent type. In our case, this is `GTypeInterface`
+#[derive(Copy, Clone, Debug)]
```

Comment:
Do we need copy/clone?

Reply:
I guess it's good to have for consistency with the autogenerated FFI structs

Reply:
The reason is apparently the `Box::new()` in `interface_init()` for the parent interface struct. For classes that's not required because we get it by pointer via `g_type_class_peek_parent()`.

 * I think we should have `Clone + Copy` on all these `repr(C)` structs for consistency with the sys crates
 * If I'm not confused, class structs must be `Copy` too because that's what GObject is going to do internally, so there's currently a little hole here. If we were exposing `base_init` this could be avoided
 * We could look at using a pointer for the interface in `interface_init()` too

Do you want to look at the second and third? If not we should create an issue at least for the second one as that seems potentially problematic :)

Reply:
Also with regards to `base_init()`, we could only require `Clone` and automatically call that in there I guess?

Reply:
Handled in https://github.com/gtk-rs/gtk-rs-core/issues/1381